src/nvim/eval.c

```diff
@@ -11558,6 +11570,529 @@ static void f_join(typval_T *argvars, typval_T *rettv)
+          EMSG2(_("E474: Missing number after decimal dot: %s"), s);
+          goto json_decode_string_fail;
+        } else if (p == exps) {
+          EMSG2(_("E474: Missing exponent: %s"), s);
+          goto json_decode_string_fail;
+        }
+        typval_T tv = {
+          .v_type = VAR_NUMBER,
+          .v_lock = VAR_UNLOCKED,
+        };
+        if (fracs) {
+          // Convert floating-point number
+          (void) string2float((char_u *) s, &tv.vval.v_float);
+          tv.v_type = VAR_FLOAT;
+        } else {
+          // Convert integer
+          long nr;
+          vim_str2nr((char_u *) s, NULL, NULL, 0, 0, 0, &nr, NULL);
+          tv.vval.v_number = (varnumber_T) nr;
+        }
+        POP(tv);
+        p--;
+        break;
+      }
+      case '[': {
+        list_T *list = list_alloc();
+        list->lv_refcount++;
+        typval_T tv = {
+          .v_type = VAR_LIST,
+          .v_lock = VAR_UNLOCKED,
+          .vval = { .v_list = list },
+        };
+        kv_push(ContainerStackItem, container_stack, ((ContainerStackItem) {
+          .stack_index = kv_size(stack),
+          .container = tv,
+        }));
+        kv_push(typval_T, stack, tv);
+        break;
+      }
+      case '{': {
+        dict_T *dict = dict_alloc();
+        dict->dv_refcount++;
+        typval_T tv = {
+          .v_type = VAR_DICT,
+          .v_lock = VAR_UNLOCKED,
+          .vval = { .v_dict = dict },
+        };
+        kv_push(ContainerStackItem, container_stack, ((ContainerStackItem) {
+          .stack_index = kv_size(stack),
+          .container = tv,
+        }));
+        kv_push(typval_T, stack, tv);
+        break;
+      }
+      default: {
+        EMSG2(_("E474: Unidentified byte: %s"), p);
+        goto json_decode_string_fail;
+      }
+    }
+    didcomma = false;
+    didcolon = false;
+    if (kv_size(container_stack) == 0) {
+      p++;
+      break;
+    }
+  }
+#undef POP
+json_decode_string_after_cycle:
+  for (; p < e; p++) {
+    switch (*p) {
+      case NL:
+      case ' ':
+      case TAB: {
+        break;
+      }
+      default: {
+        EMSG2(_("E474: Trailing characters: %s"), p);
+        goto json_decode_string_fail;
+      }
+    }
+  }
+  if (kv_size(stack) > 1 || kv_size(container_stack)) {
+    EMSG2(_("E474: Unexpected end of input: %s"), buf);
+    goto json_decode_string_fail;
+  }
+  goto json_decode_string_ret;
+json_decode_string_fail:
+  ret = FAIL;
+  while (kv_size(stack)) {
+    clear_tv(&kv_pop(stack));
+  }
+json_decode_string_ret:
+  if (ret != FAIL) {
+    assert(kv_size(stack) == 1);
+    *rettv = kv_pop(stack);
+  }
+  kv_destroy(stack);
+  kv_destroy(container_stack);
+  return ret;
+}
+
+/// jsondecode() function
+static void f_jsondecode(typval_T *argvars, typval_T *rettv)
+{
+  char numbuf[NUMBUFLEN];
+  char *s = NULL;
+  char *tofree = NULL;
+  size_t len;
+  if (argvars[0].v_type == VAR_LIST) {
+    if (!encode_vim_list_to_buf(argvars[0].vval.v_list, &len, &s)) {
```

Comment:
For input `[]` and `['']`, `json_decode_string()` returns `0` (valid JSON?).
If input is `['','']` it crashes with

```c
==13267==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x60200016ffb1 at pc 0x0000004726fe bp 0x7fff35630590 sp 0x7fff3562fd40
READ of size 2 at 0x60200016ffb1 thread T0
    #0 0x4726fd in __interceptor_strlen /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:213
    #1 0x1642a6c in vim_vsnprintf /home/oni-link/git/neovim/src/nvim/message.c:3240:25
    #2 0x164bb7d in emsgf /home/oni-link/git/neovim/src/nvim/message.c:574:3
    #3 0xeed7a0 in f_json_decode /home/oni-link/git/neovim/src/nvim/eval.c:11621:5
    #4 0xe01cd3 in call_func /home/oni-link/git/neovim/src/nvim/eval.c:7242:11
    #5 0xe15c74 in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:7091:11
    #6 0xe9e03c in eval7 /home/oni-link/git/neovim/src/nvim/eval.c:4229:15
    #7 0xe99aa5 in eval6 /home/oni-link/git/neovim/src/nvim/eval.c:3950:7
    #8 0xe96c29 in eval5 /home/oni-link/git/neovim/src/nvim/eval.c:3802:7
    #9 0xe91e06 in eval4 /home/oni-link/git/neovim/src/nvim/eval.c:3545:7
    #10 0xe91373 in eval3 /home/oni-link/git/neovim/src/nvim/eval.c:3465:7
    #11 0xe90913 in eval2 /home/oni-link/git/neovim/src/nvim/eval.c:3402:7
    #12 0xdfa7d2 in eval1 /home/oni-link/git/neovim/src/nvim/eval.c:3335:7
    #13 0xe1599f in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:7077:9
    #14 0xe9e03c in eval7 /home/oni-link/git/neovim/src/nvim/eval.c:4229:15
    #15 0xe99aa5 in eval6 /home/oni-link/git/neovim/src/nvim/eval.c:3950:7
    #16 0xe96c29 in eval5 /home/oni-link/git/neovim/src/nvim/eval.c:3802:7
    #17 0xe91e06 in eval4 /home/oni-link/git/neovim/src/nvim/eval.c:3545:7
    #18 0xe91373 in eval3 /home/oni-link/git/neovim/src/nvim/eval.c:3465:7
    #19 0xe90913 in eval2 /home/oni-link/git/neovim/src/nvim/eval.c:3402:7
    #20 0xdfa7d2 in eval1 /home/oni-link/git/neovim/src/nvim/eval.c:3335:7
    #21 0xe50e9e in ex_echo /home/oni-link/git/neovim/src/nvim/eval.c:18349:9
    #22 0x135eef0 in do_one_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:2191:5
    #23 0x133d74d in do_cmdline /home/oni-link/git/neovim/src/nvim/ex_docmd.c:601:20
    #24 0x1010a5b in nv_colon /home/oni-link/git/neovim/src/nvim/normal.c:4485:18
    #25 0xff7fb3 in normal_execute /home/oni-link/git/neovim/src/nvim/normal.c:1147:3
    #26 0x125c920 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:55:26
    #27 0xfac4c0 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #28 0xf77c06 in main /home/oni-link/git/neovim/src/nvim/main.c:538:3
    #29 0x7ff4ab3bd72f in __libc_start_main (/lib64/libc.so.6+0x2072f)
    #30 0x444408 in _start (/home/oni-link/git/neovim/build/bin/nvim+0x444408)

0x60200016ffb1 is located 0 bytes to the right of 1-byte region [0x60200016ffb0,0x60200016ffb1)
allocated by thread T0 here:
    #0 0x4e5088 in malloc /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:52
    #1 0xaf6124 in try_malloc /home/oni-link/git/neovim/src/nvim/memory.c:59:15
    #2 0xaf62e4 in xmalloc /home/oni-link/git/neovim/src/nvim/memory.c:93:15
    #3 0x177e6c8 in encode_vim_list_to_buf /home/oni-link/git/neovim/src/nvim/eval/encode.c:220:21
    #4 0xeed426 in f_json_decode /home/oni-link/git/neovim/src/nvim/eval.c:11605:10
    #5 0xe01cd3 in call_func /home/oni-link/git/neovim/src/nvim/eval.c:7242:11
    #6 0xe15c74 in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:7091:11
    #7 0xe9e03c in eval7 /home/oni-link/git/neovim/src/nvim/eval.c:4229:15
    #8 0xe99aa5 in eval6 /home/oni-link/git/neovim/src/nvim/eval.c:3950:7
    #9 0xe96c29 in eval5 /home/oni-link/git/neovim/src/nvim/eval.c:3802:7
    #10 0xe91e06 in eval4 /home/oni-link/git/neovim/src/nvim/eval.c:3545:7
    #11 0xe91373 in eval3 /home/oni-link/git/neovim/src/nvim/eval.c:3465:7
    #12 0xe90913 in eval2 /home/oni-link/git/neovim/src/nvim/eval.c:3402:7
    #13 0xdfa7d2 in eval1 /home/oni-link/git/neovim/src/nvim/eval.c:3335:7
    #14 0xe1599f in get_func_tv /home/oni-link/git/neovim/src/nvim/eval.c:7077:9
    #15 0xe9e03c in eval7 /home/oni-link/git/neovim/src/nvim/eval.c:4229:15
    #16 0xe99aa5 in eval6 /home/oni-link/git/neovim/src/nvim/eval.c:3950:7
    #17 0xe96c29 in eval5 /home/oni-link/git/neovim/src/nvim/eval.c:3802:7
    #18 0xe91e06 in eval4 /home/oni-link/git/neovim/src/nvim/eval.c:3545:7
    #19 0xe91373 in eval3 /home/oni-link/git/neovim/src/nvim/eval.c:3465:7
    #20 0xe90913 in eval2 /home/oni-link/git/neovim/src/nvim/eval.c:3402:7
    #21 0xdfa7d2 in eval1 /home/oni-link/git/neovim/src/nvim/eval.c:3335:7
    #22 0xe50e9e in ex_echo /home/oni-link/git/neovim/src/nvim/eval.c:18349:9
    #23 0x135eef0 in do_one_cmd /home/oni-link/git/neovim/src/nvim/ex_docmd.c:2191:5
    #24 0x133d74d in do_cmdline /home/oni-link/git/neovim/src/nvim/ex_docmd.c:601:20
    #25 0x1010a5b in nv_colon /home/oni-link/git/neovim/src/nvim/normal.c:4485:18
    #26 0xff7fb3 in normal_execute /home/oni-link/git/neovim/src/nvim/normal.c:1147:3
    #27 0x125c920 in state_enter /home/oni-link/git/neovim/src/nvim/state.c:55:26
    #28 0xfac4c0 in normal_enter /home/oni-link/git/neovim/src/nvim/normal.c:464:3
    #29 0xf77c06 in main /home/oni-link/git/neovim/src/nvim/main.c:538:3

SUMMARY: AddressSanitizer: heap-buffer-overflow /home/oni-link/git/llvm/projects/compiler-rt/lib/asan/../sanitizer_common/sanitizer_common_interceptors.inc:213 in __interceptor_strlen

```